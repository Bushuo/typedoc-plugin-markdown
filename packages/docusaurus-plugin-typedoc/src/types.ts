export interface PluginOptions {
  id: string;
  docsRoot: string;
  out: string;
  sidebar: SidebarOptions | null;
  readmeTitle: string | undefined;
  globalsTitle?: string | undefined;
  plugin?: string[];
  readme?: string;
  disableOutputCheck?: boolean;
  entryPoints?: any;
  entryDocument: string;
  hideInPageTOC: boolean;
  hideBreadcrumbs: boolean;
  siteDir?: string;
}

export interface FrontMatter {
  id: string;
  title: string;
  slug?: string;
  sidebar_label?: string;
  hide_title?: boolean;
}

export interface SidebarOptions {
  fullNames: boolean;
  sidebarFile: string;
  indexLabel?: string;
  readmeLabel: string;
}

export interface Sidebar {
  [sidebarId: string]: SidebarItem[];
}

export interface SidebarCategory {
  type: string;
  label: string;
  items: SidebarItem[];
}

export type SidebarItem = SidebarCategory | string;
