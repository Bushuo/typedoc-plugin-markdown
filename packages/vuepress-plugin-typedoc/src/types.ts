import { Application, ProjectReflection } from 'typedoc';

export interface PluginOptions {
  hideBreadcrumbs: boolean;
  out: string;
  sidebar?: SidebarOptions | null;
  plugin: string[];
  watch: boolean;
}

export interface SidebarOptions {
  fullNames: boolean;
  parentCategory: string;
  sidebarFile: string;
  sidebarPath: string;
}

export interface LoadedContent {
  app: Application;
  project: ProjectReflection;
}
