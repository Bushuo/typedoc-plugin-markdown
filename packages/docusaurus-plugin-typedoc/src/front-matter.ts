import * as path from 'path';

import { BindOption } from 'typedoc';
import {
  getPageTitle,
  prependYAML,
} from 'typedoc-plugin-markdown/dist/utils/front-matter';
import { Component } from 'typedoc/dist/lib/converter/components';
import { RendererComponent } from 'typedoc/dist/lib/output/components';
import { PageEvent } from 'typedoc/dist/lib/output/events';

import { FrontMatter, Sidebar } from './types';

@Component({ name: 'front-matter' })
export class FrontMatterComponent extends RendererComponent {
  @BindOption('out')
  out!: string;
  @BindOption('sidebar')
  sidebar!: Sidebar;
  @BindOption('globalsTitle')
  globalsTitle!: string;
  @BindOption('readmeTitle')
  readmeTitle!: string;
  @BindOption('entryDocument')
  entryDocument!: string;

  globalsFile = 'modules.md';

  initialize() {
    super.initialize();
    this.listenTo(this.application.renderer, {
      [PageEvent.END]: this.onPageEnd,
    });
  }
  onPageEnd(page: PageEvent) {
    if (page.contents) {
      page.contents = prependYAML(page.contents, this.getYamlItems(page));
    }
  }

  getYamlItems(page: PageEvent): any {
    const pageId = this.getId(page);
    const pageTitle = this.getTitle(page);
    const sidebarLabel = this.getSidebarLabel(page);
    let items: FrontMatter = {
      id: pageId,
      title: pageTitle,
    };
    if (page.url === this.entryDocument) {
      items = { ...items, slug: '/' + this.out };
    }
    if (sidebarLabel && sidebarLabel !== pageTitle) {
      items = { ...items, sidebar_label: sidebarLabel };
    }
    if (page.url === page.project.url) {
      items = { ...items, hide_table_of_contents: true };
    }
    return {
      ...items,
      custom_edit_url: null,
      hide_title: true,
    };
  }

  getSidebarLabel(page: PageEvent) {
    if (!this.sidebar) {
      return null;
    }
    if (page.url === this.entryDocument) {
      return page.url === page.project.url
        ? this.sidebar.indexLabel
        : this.sidebar.readmeLabel;
    }

    if (page.url === this.globalsFile) {
      return this.sidebar.indexLabel;
    }
    return this.sidebar.fullNames ? page.model.getFullName() : page.model.name;
  }

  getId(page: PageEvent) {
    return path.basename(page.url, path.extname(page.url));
  }

  getTitle(page: PageEvent) {
    const readmeTitle = this.readmeTitle || page.project.name;
    if (page.url === this.entryDocument && page.url !== page.project.url) {
      return readmeTitle;
    }
    return getPageTitle(page);
  }
}
