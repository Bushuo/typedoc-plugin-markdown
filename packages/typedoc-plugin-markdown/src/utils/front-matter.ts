import { PageEvent } from 'typedoc/dist/lib/output/events';

import { reflectionTitle } from '../resources/helpers/reflection-title';

/*
 * Front Matter variable model.
 */
export interface FrontMatterVars {
  [key: string]: string | number | boolean;
}

/*
 * Prepends YAML block to top of page.
 * @param page
 * @param vars
 */
export const addYAML = (page: PageEvent, vars: FrontMatterVars) => {
  if (page.contents) {
    page.contents = page.contents
      .replace(/^/, toYAML(vars) + '\n\n')
      .replace(/[\r\n]{3,}/g, '\n\n');
  }
};

/*
 * Returns the page title as rendered in the document h1(# title)
 * @param page
 */
export const getPageTitle = (page: PageEvent) => {
  return reflectionTitle.call(page, false);
};

/**
 * Converts YAML object to a YAML string
 * @param vars
 */
const toYAML = (vars: FrontMatterVars) => {
  const yaml = `---
${Object.entries(vars)
  .map(
    ([key, value]) =>
      `${key}: ${
        typeof value === 'string' ? `"${escapeString(value)}"` : value
      }`,
  )
  .join('\n')}
---`;
  return yaml;
};

// prettier-ignore
const escapeString=(str: string)=> {
    return str.replace(/([^\\])'/g, '$1\\\'').replace(/\"/g, '');
  }
