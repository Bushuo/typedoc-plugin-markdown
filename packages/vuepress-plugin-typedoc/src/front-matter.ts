import { reflectionTitle } from 'typedoc-plugin-markdown/dist/resources/helpers/reflection-title';
import { Component } from 'typedoc/dist/lib/converter/components';
import { RendererComponent } from 'typedoc/dist/lib/output/components';
import { PageEvent } from 'typedoc/dist/lib/output/events';

@Component({ name: 'front-matter' })
export class FrontMatterComponent extends RendererComponent {
  initialize() {
    super.initialize();
    this.listenTo(this.application.renderer, {
      [PageEvent.END]: this.onPageEnd,
    });
  }
  onPageEnd(page: PageEvent) {
    if (page.contents) {
      page.contents = page.contents
        .replace(/^/, this.getYamlString(this.getYamlItems(page)) + '\n\n')
        .replace(/[\r\n]{3,}/g, '\n\n');
    }
  }

  getYamlString(yamlItems: { [key: string]: string | number | boolean }) {
    const yaml = `---
${Object.entries(yamlItems)
  .map(
    ([key, value]) =>
      `${key}: ${
        typeof value === 'string' ? `"${this.escapeYAMLString(value)}"` : value
      }`,
  )
  .join('\n')}
---`;
    return yaml;
  }

  getYamlItems(page: PageEvent): any {
    const pageTitle = this.getTitle(page);

    return {
      title: pageTitle,
    };
  }

  getTitle(page: PageEvent) {
    return reflectionTitle.call(page, false);
  }

  // prettier-ignore
  escapeYAMLString(str: string) {
    return str.replace(/([^\\])'/g, '$1\\\'').replace(/\"/g, '');
  }
}
