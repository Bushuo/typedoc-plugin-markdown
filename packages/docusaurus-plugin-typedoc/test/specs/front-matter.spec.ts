import cuid from 'cuid';
import { PageEvent } from 'typedoc/dist/lib/output/events';

import { TestApp } from '../../../typedoc-plugin-markdown/test/test-app';
import { DocusaurusFrontMatterComponent } from '../../dist/front-matter';
import { PluginOptions } from '../../dist/types';

async function generate(testApp: TestApp, options: PluginOptions) {
  await testApp.bootstrap();
  const componentNamename = cuid();
  testApp.renderer.addComponent(
    componentNamename,
    new DocusaurusFrontMatterComponent(testApp.renderer, options),
  );
  const frontMatterComponent = testApp.renderer.getComponent(
    componentNamename,
  ) as DocusaurusFrontMatterComponent;
  return frontMatterComponent;
}

describe(`FrontMatter:`, () => {
  let testApp: TestApp;
  const DEFAULT_PLUGIN_OPTIONS: PluginOptions = {
    out: 'api',
    sidebar: {
      fullNames: false,
      sidebarFile: 'typedoc-sidebar.js',
      globalsLabel: 'Globals',
      readmeLabel: 'README',
    },
  } as PluginOptions;
  beforeAll(() => {
    testApp = new TestApp(['frontmatter.ts']);
  });

  describe(`(readme)`, async () => {
    let page: PageEvent;

    beforeEach(() => {
      page = {
        url: 'index.md',
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
    });

    test(`should set default index page`, async () => {
      const frontMatterComponent = await generate(
        testApp,
        DEFAULT_PLUGIN_OPTIONS,
      );
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set custom readmeLabel and readmeTitle`, async () => {
      const frontMatterComponent = await generate(testApp, {
        ...DEFAULT_PLUGIN_OPTIONS,
        sidebar: {
          ...DEFAULT_PLUGIN_OPTIONS.sidebar,
          readmeLabel: 'Custom readme label',
        },
        readmeTitle: 'Custom readme title',
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set the same custom readmeLabel and readmeTitle`, async () => {
      const frontMatterComponent = await generate(testApp, {
        ...DEFAULT_PLUGIN_OPTIONS,
        sidebar: {
          ...DEFAULT_PLUGIN_OPTIONS.sidebar,
          readmeLabel: 'Custom readme title',
        },
        readmeTitle: 'Custom readme title',
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set default globals page`, async () => {
      const frontMatterComponent = await generate(testApp, {
        ...DEFAULT_PLUGIN_OPTIONS,
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });
  });

  describe(`(globals)`, () => {
    let page: PageEvent;

    beforeEach(() => {
      page = {
        url: 'modules.md',
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
    });

    test(`should set default globals page`, async () => {
      const frontMatterComponent = await generate(
        testApp,
        DEFAULT_PLUGIN_OPTIONS,
      );
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set custom globalsLabel and custom globalsTitle`, async () => {
      const frontMatterComponent = await generate(testApp, {
        ...DEFAULT_PLUGIN_OPTIONS,
        sidebar: {
          ...DEFAULT_PLUGIN_OPTIONS.sidebar,
          globalsLabel: 'Custom globals label',
        },
        globalsTitle: 'Custom globals title',
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should set globals page when readme=none`, async () => {
      page.url = 'index.md';
      page.project.url = 'index.md';
      const frontMatterComponent = await generate(
        testApp,
        DEFAULT_PLUGIN_OPTIONS,
      );
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });
  });

  describe(`(reflection names)`, () => {
    let page: PageEvent;

    beforeEach(() => {
      const reflection = testApp.findReflection('FrontMatterClass');

      page = {
        project: testApp.project,
        model: reflection,
        url: 'url',
        contents: 'CONTENTS',
      } as PageEvent;
    });

    test(`should return reflection labels with short names`, async () => {
      const frontMatterComponent = await generate(
        testApp,
        DEFAULT_PLUGIN_OPTIONS,
      );
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });

    test(`should return front matter with full names`, async () => {
      const frontMatterComponent = await generate(testApp, {
        ...DEFAULT_PLUGIN_OPTIONS,
        sidebar: {
          ...DEFAULT_PLUGIN_OPTIONS.sidebar,
          fullNames: true,
        },
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });
  });

  describe(`(no sidebar)`, () => {
    test(`should return front matter without sidebar`, async () => {
      const page = {
        url: 'modules.md',
        project: { name: 'test-project-name', url: 'modules.md' },
        contents: 'CONTENTS',
      } as PageEvent;
      const frontMatterComponent = await generate(testApp, {
        ...DEFAULT_PLUGIN_OPTIONS,
        sidebar: null,
      });
      frontMatterComponent.onPageEnd(page);
      expect(page.contents).toMatchSnapshot();
    });
  });
});
