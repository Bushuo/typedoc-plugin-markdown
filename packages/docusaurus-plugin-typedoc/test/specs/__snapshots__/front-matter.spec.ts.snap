// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FrontMatter: (globals) should set custom indexLabel 1`] = `
"---
id: \\"modules\\"
title: \\"test-project-name\\"
sidebar_label: \\"Custom index label\\"
hide_title: true
---

CONTENTS"
`;

exports[`FrontMatter: (globals) should set default globals page 1`] = `
"---
id: \\"modules\\"
title: \\"test-project-name\\"
sidebar_label: \\"Table of contents\\"
hide_title: true
---

CONTENTS"
`;

exports[`FrontMatter: (globals) should set globals page when readme=none 1`] = `
"---
id: \\"index\\"
title: \\"test-project-name\\"
slug: \\"/api\\"
sidebar_label: \\"Table of contents\\"
hide_title: true
---

CONTENTS"
`;

exports[`FrontMatter: (no sidebar) should return front matter without sidebar 1`] = `
"---
id: \\"modules\\"
title: \\"test-project-name\\"
hide_title: true
---

CONTENTS"
`;

exports[`FrontMatter: (readme) should set custom readmeLabel and readmeTitle 1`] = `
"---
id: \\"index\\"
title: \\"Custom readme title\\"
slug: \\"/api\\"
sidebar_label: \\"Custom readme label\\"
hide_title: true
---

CONTENTS"
`;

exports[`FrontMatter: (readme) should set default index page 1`] = `
"---
id: \\"index\\"
title: \\"test-project-name\\"
slug: \\"/api\\"
sidebar_label: \\"Readme\\"
hide_title: true
---

CONTENTS"
`;

exports[`FrontMatter: (readme) should set the same custom readmeLabel and readmeTitle 1`] = `
"---
id: \\"index\\"
title: \\"Custom readme title\\"
slug: \\"/api\\"
hide_title: true
---

CONTENTS"
`;

exports[`FrontMatter: (reflection names) should return front matter with full names 1`] = `
"---
id: \\"url\\"
title: \\"Class: FrontMatterClass\\"
sidebar_label: \\"frontmatter.FrontMatterClass\\"
hide_title: true
---

CONTENTS"
`;

exports[`FrontMatter: (reflection names) should return reflection labels with short names 1`] = `
"---
id: \\"url\\"
title: \\"Class: FrontMatterClass\\"
sidebar_label: \\"FrontMatterClass\\"
hide_title: true
---

CONTENTS"
`;
